keps/sig-node/2411-cri-container-log-rotation/README.md

```diff
@@ -0,0 +1,301 @@
+
+### Goals
+
+- The kubelet assigns the log path for a container and runtime writes the container output to that path
+- The kubelet periodically checks the disk space occupied by the container logs and rotates them if necessary
+- After the logs are rotated, the kubelet sends a signal to the container runtime to re-open the log file
+- The kubelet exposes a consistent log directory structure with metadata so that any logging agent can integrate with it
+
+### Non-Goals
+
+- Shipping the logs directly to a remote storage
+- Supporting container runtimes that run on a different virtual/physical machine from the kubelet
+
+## Proposal
+
+Graduate the CRIContainerLogRotation feature gate from Beta to Stable. The kubelet already decides the container log directory structure and passes that down to the container runtime. The container runtime then writes the container's logs to this location. This makes the kubelet the best candidate to manage the rotation of the container logs as it already know the log directory structure and has access to it. The CRILogRotation feature gate implementation adds a container log manager package, which manages and rotates the container logs. It also adds 2 flags to the kubelet that allows the user to configure the maximum size of each log file and the maximum number of log files to retain. These flags are **--container-log-max-size** and **--container-log-max-files**.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a kubernetes user, I want to use a CRI container runtime and want the container logs to be managed and rotated by the kubelet.
+
+#### Story 2
+
+As a kubernetes user, I want to integrate a logging agent that aggregates the logs to a remote storage without having to further integrate with the CRI container runtime.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+- Loss of some logs during log rotation. There is an open issue on this with a suggested fix https://github.com/kubernetes/kubernetes/issues/64760
+
+## Design Details
+
+This implementation adds container log manager package, which the kubelet uses to manage and rotate the logs. The container log manager will only start up when the container runtime being used is one that communicates with the kubelet via the CRI i.e CRI-O, containerd, etc.
+
+The rotated logs are compressed with gzip. The latest rotated log is not compressed as a logging agent, such as fluentd, might still be reading it right after rotation. The kubelet periodically checks the amount of disk space being used by the container logs and rotates them if the max value has been reached. After the logs are rotated, the kubelet sends a signal to the container runtime to re-open the log file.
```

Comment:
Comment for the first sentence - it may be both, logging agent may be reading that file and container runtime is still writing to that file shortly after the new file path was passed to the container runtime.

I wonder if container runtime implementation should forbid it from writing to older files or not. If so - how long does the runtime has to empty logging buffers (if any) before switching to the new file?

Reply:
Fixed the first sentence.

Reply:
@mrunalp wdyt about the runtime restricting where the logs are written after it gets a new log file path?

Reply:
The runtime is supposed to re-open the log when it receives the CRI call. There is no timeout specified (beyond the CRI timeout) for that.